File: nodejs/src/example-api-integration/hello-server-protected.js
// @link https://approov.io/docs/v2.1/approov-usage-documentation/#example-api-integration
// @link https://github.com/approov/documentation_backend-code-snippets/blob/master/nodejs/src/example-api-integration/hello-server-protected.js

const express = require('express')
const jwt = require('express-jwt')
const app = express()

const dotenv = require('dotenv').config()

if (dotenv.error) {
  console.debug('FAILED TO PARSE `.env` FILE | ' + dotenv.error)
}

const SECRET = dotenv.parsed.APPROOV_BASE64_SECRET

const ERROR_RESPONSE_BODY = {}

// Callback that performs the Approov token check using the express-jwt library
const checkApproovToken = jwt({
  secret: Buffer.from(SECRET, 'base64'), // decodes the Approov secret
  requestProperty: 'approovTokenDecoded',
  getToken: function fromApproovTokenHeader(req) {
    req.approovTokenError = false
    return req.get('Approov-Token')
  },
  algorithms: ['HS256']
})

// Callback to handle the errors occurred while checking the Approov token.
const handlesApproovTokenError = function(err, req, res, next) {
  if (err.name === 'UnauthorizedError') {
    req.approovTokenError = true
    res.status(401)
    res.json(ERROR_RESPONSE_BODY)
    return
  }

  next()
  return
}

// Callback to handles when an Approov token is successfully validated.
const handlesApproovTokenSuccess = function(req, res, next) {
  if (req.approovTokenError === true) {
    res.status(401)
    res.json(ERROR_RESPONSE_BODY)
    return
  }

  next()
  return
}

// Intercepts all calls to the 'hello world' endpoint and validates the Approov token.
app.use('/', checkApproovToken)

// Handles failure in validating the Approov token
app.use('/', handlesApproovTokenError)

// Handles requests where the Approov token is a valid one.
app.use('/', handlesApproovTokenSuccess)

// simple 'hello world' endpoint.
app.get('/', function (req, res, next) {
    res.json({message: "Hello Approov token protected server!"})
})

// Create and run the HTTP server
app.listen(8002, function () {
  console.debug("Server listening on %s", "http://localhost:8002")
})

File: python/src/example-api-integration/hello-server-protected.py
# @link https://approov.io/docs/v2.0/approov-usage-documentation/#example-api-integration
# @link https://github.com/approov/documentation_backend-code-snippets/blob/master/python/src/example-api-integration/hello-server-protected.py

from flask import Flask, request, abort, jsonify, make_response
import jwt # https://github.com/jpadilla/pyjwt/
import json
import base64

from os import getenv
from dotenv import load_dotenv, find_dotenv
load_dotenv(find_dotenv(), override=True)

# Token secret value obtained with the Approov CLI tool: approov secret <admin.tok> -get
SECRET = getenv('APPROOV_BASE64_SECRET')

app = Flask(__name__)

def verifyApproovToken(request):
  # Get the Approov Token from header
  approov_token = request.headers.get("Approov-Token")

  # If we didn't find a token, then reject the request
  if approov_token == "":
    return None

  try:
    # Decode our token, allowing only the HS256 algorithm, using our base64 encoded SECRET
    approov_token_claims = jwt.decode(approov_token, base64.b64decode(SECRET), algorithms=['HS256'])
  except jwt.ExpiredSignatureError as e:
    # Signature has expired, token is bad
    return None
  except jwt.InvalidTokenError as e:
    # Token could not be decoded, token is bad
    return None

  return approov_token_claims

@app.route("/")
def hello():
  approov_token_claims = verifyApproovToken(request)

  if approov_token_claims is None:
    abort(make_response(jsonify({}), 401))

  return jsonify({"message": "Hello World!"})

if __name__ == "__main__":
   app.run()
